Sources/Algorithms/Permutations.swift

```diff
@@ -82,72 +113,95 @@ extension Permutations: Sequence {
     /// - Complexity: O(*n*), where *n* is the length of the collection.
     @usableFromInline
     internal mutating func nextState() -> Bool {
+      let countToChoose = self.kRange.lowerBound
       let edge = countToChoose - 1
-
+      
       // Find first index greater than the one at `edge`.
       if let i = indexes[countToChoose...].firstIndex(where: { indexes[edge] < $0 }) {
         indexes.swapAt(edge, i)
       } else {
-        indexes.reverse(subrange: countToChoose..<indexes.endIndex)
-
+        indexes.reverse(subrange: countToChoose ..< indexes.endIndex)
+        
         // Find last increasing pair below `edge`.
         // TODO: This could be indexes[..<edge].adjacentPairs().lastIndex(where: ...)
         var lastAscent = edge - 1
         while (lastAscent >= 0 && indexes[lastAscent] >= indexes[lastAscent + 1]) {
           lastAscent -= 1
         }
-        if (lastAscent < 0) {
+        if lastAscent < 0 {
           return false
         }
-
+        
         // Find rightmost index less than that at `lastAscent`.
         if let i = indexes[lastAscent...].lastIndex(where: { indexes[lastAscent] < $0 }) {
           indexes.swapAt(lastAscent, i)
         }
-        indexes.reverse(subrange: (lastAscent + 1)..<indexes.endIndex)
+        indexes.reverse(subrange: (lastAscent + 1) ..< indexes.endIndex)
       }
       
       return true
     }
     
     @inlinable
     public mutating func next() -> [Base.Element]? {
-      if !hasMorePermutations { return nil }
+      guard !isFinished else { return nil }
       
+      /// Advances `kRange` by incrementing its `lowerBound` until the range is
+      /// empty, when iteration is finished.
+      func advanceKRange() {
+        if kRange.lowerBound < kRange.upperBound {
+          let advancedLowerBound = kRange.lowerBound + 1
+          kRange = advancedLowerBound ..< kRange.upperBound
+          indexes = Array(base.indices)
+        }
```

Comment:
```suggestion
        kRange.removeFirst()
        indexes = Array(base.indices)
```

You've already checked that `!isFinished` (i.e. `!kRange.isEmpty`), so no need to check that here.

Reply:
Good catch. Thank you!